internal/pkg/log/log.go

```diff
@@ -2,105 +2,123 @@
-type Options struct {
-	// Level configures the log verbosity. Defaults to Debug.
-	Level *zap.AtomicLevel
-	// StacktraceLevel is the level which stacktraces will be emitted. Defaults
-	// to Warn.
-	StacktraceLevel *zap.AtomicLevel
-	// Encoder configures how Zap will encode the output. Defaults to JSON.
-	Encoder zapcore.Encoder
-	// OutputDest controls the destination of the log output. Defaults to
-	// os.Stderr.
-	OutputDest io.Writer
-	// ZapOptions allows passing additional optional zap.Options, ex: Sampling.
-	ZapOptions []zap.Option
+// Logger is the contract for xds-relay's logging implementation.
+//
+// A self-contained usage example looks as follows:
+//
+//    Log.Named("foo-component").With(
+//        "field1", "value1",
+//        "field2", "value2",
+//     ).Error("my error message")
+type Logger interface {
+	// Named adds a sub-scope to the logger.
+	Named(name string) Logger
+
+	// With adds a variadic number of fields to the logging context.
+	// When processing pairs, the first element of the pair is used as the
+	// field key and the second as the field value.
+	//
+	// For example,
+	//
+	//   Log.With(
+	//     "hello", "world",
+	//     "failure", errors.New("oh no"),
+	//     "count", 42,
+	//     "user", User{Name: "alice"},
+	//  ).Info("this is an error message")
+	With(args ...interface{}) Logger
+
+	// Log a message at level Debug, annotated with fields provided through With().
+	Debug(ctx context.Context, msg ...interface{})
+
+	// Log a message at level Info, annotated with fields provided through With().
+	Info(ctx context.Context, msg ...interface{})
+
+	// Log a message at level Warn, annotated with fields provided through With().
+	Warn(ctx context.Context, msg ...interface{})
+
+	// Log a message at level Error, annotated with fields provided through With().
+	Error(ctx context.Context, msg ...interface{})
+
+	// Log a message at level Panic, annotated with fields provided through With(), and immediately
+	// panic.
+	Panic(ctx context.Context, msg ...interface{})
+
+	// Log a message at level Fatal, annotated with fields provided through With(), and immediately
+	// call os.Exit.
+	Fatal(ctx context.Context, msg ...interface{})
+
+	// Sync flushes any buffered log entries.
+	Sync() error
 }
 
-// Opts allows manipulation of the Zap options.
-type Opts func(*Options)
+type logger struct {
+	zap *z.SugaredLogger
+}
 
-// addDefaults adds defaults to the Options
-func (o *Options) addDefaults() {
-	if o.OutputDest == nil {
-		o.OutputDest = os.Stderr
-	}
-	if o.Encoder == nil {
-		encCfg := zap.NewProductionEncoderConfig()
-		o.Encoder = zapcore.NewJSONEncoder(encCfg)
-	}
-	if o.Level == nil {
-		level := zap.NewAtomicLevelAt(zap.DebugLevel)
-		o.Level = &level
-	}
-	if o.StacktraceLevel == nil {
-		level := zap.NewAtomicLevelAt(zap.WarnLevel)
-		o.StacktraceLevel = &level
+// New returns an instance of Logger implemented using the Zap logging framework.
+func New(logLevel string) Logger {
```

Comment:
I think we should move this to a separate go file and keep the interface standalone with no dependencies to the zap framework. But this is just a suggestion.